main.py

```diff
@@ -114,30 +162,12 @@ def main(args):
     if args.print_and_quit:
         print(expanded_bpf_text)
         sys.exit(0)
+    out = utils.smart_open(args.output_file, mode='w')
     b = BPF(text=expanded_bpf_text)
-    with utils.smart_open(args.output_file, mode='w') as out:
-        while True:
-            trace = b.trace_readline()
-            # print(trace)
-            proctree_enriched = []
-            error = ""
-            try:
-                # FIXME: this next line isn't right - sometimes there are more colons
-                json_event = trace.split(":", 2)[2:][0]
-                event = json.loads(json_event)
-                proc = psutil.Process(event["pid"])
-                proctree = utils.crawl_process_tree(proc)
-                proctree_enriched = list({"pid": p.pid, "cmdline": " ".join(p.cmdline()), "username":  p.username()} for p in proctree)
-            except Exception as e:
-                error=str(e)
-            print >> out, json.dumps(
-                {"timestamp": datetime.utcnow().isoformat() + 'Z',
-                 "pid": event["pid"],
-                 "proctree": proctree_enriched,
-                 "daddr": socket.inet_ntoa(struct.pack('<L', int(event["daddr"], 16))),
-                 "port": event["dport"],
-                 "error": error})
-            out.flush()
+    b["events"].open_perf_buffer(partial(print_enriched_event, b, out))
+    while True:
+        b.perf_buffer_poll()
```

Comment:
This is an infinite loop, so the two lines below will never be called. Is that intentional?

Reply:
Yeah, it's always been like that because I was going to wrap it in a `try/finally` in order to catch various signals and close gracefully, the `out.close()` and `sys.exit()` are kind of vestigial reminders to do that.

Will cut an issue to implement the signal handling :)

Reply:
#9 logged for this